### ISSUE TITLE
Introduce `Heap[r]` effect

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#8910]
I would like to introduce a `Heap[r]` effect. It would look like this:

```scala
pub def range(rc: Region[r], b: Int32, e: Int32): Array[Int32, r] \ Heap[r] =
```

Intuitively, `r` has kind `Region` and `Heap` is a type constructor: `Heap: Region -> Eff`

**Comment by mlutze** [comment]
- Introduce new kind `Region`
- Introduce `RegionUnification` -- presumably just identity
- Copy or generalize the treatment of associated effects in EffUnificationX to work for heaps. They're mostly treated as rigid variables
- Probably introduce `RegionSym` to avoid the whole "special rigid variable" thing
  - this means we need to reimplement stuff for purifying regions (since that used substitution before)

There will be less to do if we merge my new constraint solver PR first.

**Comment by magnus-madsen** [comment]
> There will be less to do if we merge my new constraint solver PR first.

Definitely, I am not planning to this before the new stuff is in. I hope to start reviewing your code soon.

Meanwhile, we can start refactoring with:

```
type alias Heap[r: Eff] = r
```

**Comment by mlutze** [comment]
> Meanwhile, we can start refactoring with:
> 
> ```
> type alias Heap[r: Eff] = r
> ```

I bet this will shine a light on many bugs

**Comment by JonathanStarup** [comment]
After discussion with Matt, i think that we should change it like so:

- We add a kind `Region` that replaces `Eff` (e.g. `Array: * -> Region -> *`)
- We add a type constructor `Heap` that has kind `Region -> Eff`
- We add `TypeConstructor.Region(sym: Symbol.RegionSym, loc)` with kind `Region`
- This now implies that we have variables of kind `Region`, both flexible and rigid. so `region rc` is a concrete region, a formal parameter might have rigid region variables, and inference might find flexible region variables.
- Now, in our effects we have `Heap[flex]`, `Heap[rigid]`, and `Heap[concrete]`, How do we convert them into the set algebra?
    - `Heap[concrete]` is a concrete set element with known identity, hence it becomes a set element, `e42`
    - `Heap[rigid]` is more tricky. We dont know whether `Heap[r1]` is disjoint from `Heap[r2]` so they cannot be elements. But we also know that `Heap[r1]` is not `Crash + IO` fx. We know that `exists rc: Region. r1 = rc`. We can gain part of this by saying that `Heap[rigid]` is a constant, representing a set element. This is different from our regular constants that represent an arbitrary set.
    - `Heap[flex]` is among the same lines a variable that represents an unknown element instead of an unknown set.
- Based on the above points, we add new `SetFormula` variables and constants that represent elements and not sets. how do we handle them during solving? 🤷
- One part i skipped over in my translation of `exists rc: Region. r1 = rc` is that we know that `Heap[r]` is a region element, and not just any element like `Crash`. The above steps does not capture that. I think we need to solve that with effect subsets such as `AE` like we've discussed before.

**Comment by magnus-madsen** [comment]
I agree with the first steps. Its the last steps where it becomes tricky.

**Comment by magnus-madsen** [comment]
Many thanks for the write-up! It helped me improve my thinking. I now think that:

- I agree that `Heap[concrete]` is just an element. Easy case.
- I think `Heap[rigid]` is just an element. In particular, it can **only** be equal to itself. If we have `Heap[r1]` and `Heap[r2]` where __both__ are rigid then we know they are disjoint. Or perhaps stated slightly differently: They can be the same, but we are not allowed to make that assumption. For a simple example, consider `def foo(x: a, y: b)`. Both `a` and `b` are rigid in the body. They can be the same, maybe they are both strings, but we are not allowed to assume that. 
- I think `Heap[flex]` is the tricky case. For example, what if we have some equation like `Heap[flex] + Heap[r17] = Heap[r18] + Heap[r17]`?  I am not sure how to solve that (note: this example has multiple solutions). How can such a constraint arise, I think from something like: `x -> x[0]` and then a usage of the lambda.

**Comment by JonathanStarup** [comment]
> If we have Heap[r1] and Heap[r2] where **both** are rigid then we know they are disjoint.

No? Any given rigid variables can be the same, whether it's a type/effect/region. A constant means that we know nothing. Concrete elements mean that we have extra (unsound) assumptions.

The assumption that they are disjoint (concrete element) would make the following program inexpressible (it makes set operations unsound)
```
/// Requires two functions of disjoint effects.
def parRun(f: Unit -> Unit \ ef, g: Unit -> Unit \ any - ef): Unit \ ef + (any - ef) = ???

def incInPar(a1: Array[Int32, r1], a2: Array[Int32, r2]): Unit \ r1 + r2 =
    let f1: _ -> _ \ r1 = () -> Array.transform!(Add.add(1), a1);
    let f2: _ -> _ \ r2 = () -> Array.transform!(Add.add(1), a2);
    parRun(f1, f2) // oh no! a race!
```

**Comment by magnus-madsen** [comment]
Does Flix handle your example correctly today?

**Comment by JonathanStarup** [comment]
yeah, because `r1` and `r2` are constants, `r2` cannot unify with `r2 - r1` in the call to parRun
